### PR TITLE
Wakelock api v2

### PR DESCRIPTION
**Pull request by hmallat**

### CODE REVIEW

File: bluez/audio/main.c [hunk 1]
```diff
@@ -45,9 +45,40 @@
...
+							audio_wakelock_put,
+							NULL);
+	if (audio_wakelock_source)
+		wakelock_acquire(audio_wakelock);
+
+	if (old_source) {
+		g_source_remove(old_source);
+		wakelock_release(audio_wakelock);
+	}
+}
```
Comment (spiiroin) [inline bluez/audio/main.c]: This looks a bit odd. I'm assuming the intent is to take wakelock on 1st call & install timer; renew the timer on subsequent calls. Now AFAICT audio_wakelock_get() x 2 -> timer active, but wakelock released.

Perhaps something like this?

if( !audio_wakelock_source )
  wakelock_acquire(audio_wakelock);
else
  g_source_remove(audio_wakelock_source);
audio_wakelock_source = g_timeout_add_seconds(...)

(plus release on glib timer failure, if that was the intention)

File: bluez/plugins/jolla-wakelock.c [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+static gboolean real_lock_acquired = FALSE;
+
+static int sysfs_write(const gchar *file, const gchar *data_ptr, gsize data_len)
+{
+	int fd = -1;
+	int r;
+
+	fd = open(file, O_WRONLY);
+	if (fd < 0) {
+		error("Failed to open: %s(%d)", strerror(errno), errno);
```
Comment (spiiroin) [inline bluez/plugins/jolla-wakelock.c]: Not sure if it matters here, but strerror() is not thread safe. If error() can be assumed to go to glibc formatter, them "%m" could be used instead of "%s" + strerror(errno)

File: bluez/audio/main.c (hunk 1 above)
Comment (hmallat) [inline bluez/audio/main.c]: audio_wakelock has reference count, so the intent is like this:

- new timer setup ok -> increment reference count
- clean up old timer -> decrement reference count

File: bluez/plugins/jolla-wakelock.c [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+		struct wakelock *check = l->data;
+		if (check->acquisitions)
+			goto done;
+	}
+
+	r = sysfs_write(UNLOCK_FILE, LOCK_NAME, LOCK_NAME_LEN);
+	if (r < 0) {
+		error("Failed to release real lock for lock: %p, name: %s",
+			lock, lock->name);
+		return r;
```
Comment (spiiroin) [inline bluez/plugins/jolla-wakelock.c]: Not sure how real the threat is, but if somebody else removes the "bluetoothd" wakelock, this error path gets hit
and book keeping stays at "have real lock" state incorrectly. Just might be lesser evil to log the error, but not bail out via early return so that the next acquire at least attempts to take the lock again?

File: bluez/plugins/jolla-wakelock.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+							NULL);
+		g_key_file_free(config);
+	}
+
+	if (!use_wakelocks) /* No need to initialize or uninitialize */
+		return 0;
+
+	/* Sanity: release any left over wakelock from previous processes
+	   (e.g. after a crash-restart) */
+	sysfs_write(UNLOCK_FILE, LOCK_NAME, LOCK_NAME_LEN);
```
Comment (spiiroin) [inline bluez/plugins/jolla-wakelock.c]: If there is no lock to cleanup, sysfs_write() can hit EINVAL on write & possibly spam the journal - might be nice to avoid that.

File: bluez/plugins/jolla-wakelock.c (hunk 3 above)
Comment (hmallat) [inline bluez/plugins/jolla-wakelock.c]: Good point, should at least test errno I guess.

File: bluez/audio/main.c (hunk 1 above)
Comment (spiiroin) [inline bluez/audio/main.c]: ok, missed that - now it makes sense

**Comment by spiiroin** [comment]
Couple of minor corner case remarks, otherwise LGTM

File: bluez/plugins/jolla-wakelock.c (hunk 2 above)
Comment (hmallat) [inline bluez/plugins/jolla-wakelock.c]: bluetoothd is single-threaded (some forking but no threads) and uses strerror() throughout, so I won't change this but good to keep in mind.

File: bluez/plugins/jolla-wakelock.c (hunk 4 above)
Comment (hmallat) [inline bluez/plugins/jolla-wakelock.c]: True, could add a silence flag there
